add natural key example

mirnawong1 opened the pull request:
this pr adds a natural key example scenario in the entities doc . raised by @patkearns10 and adding to the docs for clarity.

[slack thread](https://dbt-labs.slack.com/archives/C03KHQRQUBX/p1713924057397259)

mirnawong1 committed:
add natural key example

this pr adds a natural key example scenario in the entities doc . raised by @patkearns10 and adding to the docs for clarity.

[slack thread](https://dbt-labs.slack.com/archives/C03KHQRQUBX/p1713924057397259)

mirnawong1 committed:
Update entities.md

mirnawong1 committed:
clarify scd table and key for scd table

clarifying scd table and the required key for it

raised in [slack discussion](https://dbt-labs.slack.com/archives/C03KHQRQUBX/p1713997361959969?thread_ts=1713924057.397259&cid=C03KHQRQUBX)

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Merge branch 'mirnawong1-patch-26' into mirnawong1-patch-12

mirnawong1 committed:
Update dimensions.md

mirnawong1 committed:
Update website/docs/docs/build/dimensions.md

mirnawong1 committed:
Update website/docs/docs/build/dimensions.md

rpourzand commented on website/docs/docs/build/entities.md:
```diff
@@ -54,3 +55,15 @@ entities:
 
+### Combine columns with a key
+
+If there's a scenario where a table doesn't have a primary key, use the 'surrogate combination' to combine two columns as a key, helping you uniquely identify a record. This applies to any [entity type](/docs//build/entities#entity-types). For example, you can combine `date_key` and `brand_code` from the `raw_brand_target_weekly` table to form a `surrogate key`. The following example creates a `s ...
```
It doesn't have to be "primary"! could be any key

patkearns10 replied:
In the verbal example, it talks about not having a primary key, so I think for that example, it should be primary key. Here is what I ended up using for my most recent build:
```
# -----------------------
# --- semantic model ---
  - name: sem_department_target_weekly
    model: ref('fact_department_target_weekly')
    description: Describes weekly sales target for department for the current year
    defaults:
      agg_time_dimension: date_field

# --- entities ---
    entities:
      - name: sem_department_target_weekly_pk
        type: primary
        expr: date_key || '|' || department_code
      - name: date_key
        type: foreign
      - name: department_code
        type: foreign
```

mirnawong1 replied:
thanks @rpourzand and @patkearns10 - i've taken both of your feedback and changed it to the below. let me know what you think:

> If there's a scenario where a table doesn't have any key, such as a primary key, use the 'surrogate combination' to combine two columns as a key to help you identify a record. This applies to any [entity type](/docs//build/entities#entity-types)...

rpourzand reviewed (dismissed):
looks good just one update

mirnawong1 committed:
clarify scd table and key for scd table (#5357)

clarifying scd table and the required key for it

raised in [slack
discussion](https://dbt-labs.slack.com/archives/C03KHQRQUBX/p1713997361959969?thread_ts=1713924057.397259&cid=C03KHQRQUBX)

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Update entities.md

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

nghi-ly commented on website/docs/docs/build/dimensions.md:
```diff
@@ -24,7 +24,7 @@ All dimensions require a `name`, `type` and in some cases, an `expr` parameter.
 | `type_params` | Specific type params such as if the time is primary or used as a partition | Required |
 | `description` | A clear description of the dimension | Optional |
-| `expr` | Defines the underlying column or SQL query for a dimension. If no `expr` is specified, MetricFlow will use the column with the same name as the group. You can use column name itself to input a SQL expression. | Optional |
+| `expr` | Defines the underlying column or SQL query for a dimension. If no `expr` is specified, MetricFlow will use the column with the same name as the group. You can use the column name itself to input a SQL expression. | Optional |
```
```suggestion
| `expr` | Defines the underlying column or SQL query for a dimension. If no `expr` is specified, MetricFlow will use the column with the same name as the group. You can use the column name to input a SQL expression. | Optional |
```

nghi-ly commented on website/docs/docs/build/dimensions.md:
```diff
@@ -208,9 +208,15 @@ measures:
-As their name suggests SCD Type II are groups that change values at a coarser time granularity. This results in a range of valid rows with different dimensions values for a given metric or measure. MetricFlow associates the metric with the first (minimum) available dimensions value within a coarser time window, such as month. By default, MetricFlow uses the group that is valid at the beginning of ...
+As their name suggests SCD Type II are groups that change values at a coarser time granularity. This results in a range of valid rows with different dimensions values for a given metric or measure. MetricFlow associates the metric with the first (minimum) available dimensions value within a coarser time window, such as a month. By default, MetricFlow uses the group that is valid at the beginning  ...
+
+**SCD tables and keys**
```
sugg: promote to level H4 heading since it reads and looks like a heading

```suggestion
#### SCD tables and keys
```

nghi-ly commented on website/docs/docs/build/dimensions.md:
```diff
@@ -208,9 +208,15 @@ measures:
+
+**SCD tables and keys**
+
+SCD Type II tables have a particular dimension with a start and end date.  To join tables, use the additional [entity `type`](/docs/build/entities#entity-types) parameter, `natural` key. Using a `natural` key as an [entity `type`](/docs/build/entities#entity-types) means you also don't usually need a `primary` key. In most instances, SCD tables don't have a logically usable `primary` key because  ...
```
```suggestion
SCD Type II tables have a specific dimension with a start and end date. To join tables, set the additional [entity `type`](/docs/build/entities#entity-types) parameter to the `natural` key. Using a `natural` key as an [entity `type`](/docs/build/entities#entity-types) means you also don't usually need a `primary` key. In most instances, SCD tables don't have a logically usable `primary` key because `natural` keys map to multiple rows.
```

nghi-ly commented on website/docs/docs/build/dimensions.md:
```diff
@@ -208,9 +208,15 @@ measures:
+
+SCD Type II tables have a particular dimension with a start and end date.  To join tables, use the additional [entity `type`](/docs/build/entities#entity-types) parameter, `natural` key. Using a `natural` key as an [entity `type`](/docs/build/entities#entity-types) means you also don't usually need a `primary` key. In most instances, SCD tables don't have a logically usable `primary` key because  ...
+
+**Structure of an SCD Type II table**
```
```suggestion
#### Structure of an SCD Type II table
```

nghi-ly commented on website/docs/docs/build/entities.md:
```diff
@@ -14,8 +14,8 @@ Entities can be specified with a single column or multiple columns. Entities (jo
-:::tip Use entities as a dimensions
-You can also use entities as a dimensions, which allows you to aggregate a metric to the granularity of that entity.
+:::tip Use entities as dimensions
+You can also use entities as dimensions, which allows you to aggregate a metric to the granularity of that entity.
```
nice catch!

nghi-ly commented on website/docs/docs/build/entities.md:
```diff
@@ -26,16 +26,17 @@ MetricFlow's join logic depends on the entity `type` you use, and it also determ
+    type: Primary or natural or foreign or unique ## Required
+    description: A description of the field or role the entity takes in this table ## Optional
+    expr: The field that denotes that entity (transaction_id).  ## Optional
+          If not specified, it will default to name 
```
```suggestion
          Defaults to name if unspecified.
```

nghi-ly commented on website/docs/docs/build/entities.md:
```diff
@@ -54,3 +55,15 @@ entities:
 
+### Combine columns with a key
+
+If there's a scenario where a table doesn't have any key, such as a primary key, use the 'surrogate combination' to combine two columns as a key to help you identify a record. This applies to any [entity type](/docs//build/entities#entity-types). For example, you can combine `date_key` and `brand_code` from the `raw_brand_target_weekly` table to form a `surrogate key`. The following example creat ...
```
sugg: use italics for emphasis 

```suggestion
If a table doesn't have any key (like a primary key), use _surrogate combination_ to form a key that will help you identify a record by combining two columns. This applies to any [entity type](/docs//build/entities#entity-types). For example, you can combine `date_key` and `brand_code` from the `raw_brand_target_weekly` table to form a _surrogate key_. The following example creates a surrogate key by joining `date_key` and `brand_code` using a pipe (`|`) as a separator. 
```

nghi-ly reviewed (commented):
looks great! left some minor suggs

mirnawong1 committed:
Update website/docs/docs/build/dimensions.md

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/docs/build/dimensions.md

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/docs/build/dimensions.md

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/docs/build/entities.md

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/docs/build/dimensions.md

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/docs/build/entities.md

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

nghi-ly commented:
looks good, @mirnawong1 , from the Docs side of things!

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-26

mirnawong1 commented:
preview looks good and going to merge: https://docs-getdbt-qwmik54o8-dbt-labs.vercel.app/docs/build/dimensions#scd-type-ii
